### PR TITLE
Comments property

### PR DESCRIPTION
**Pull request by MasseGuillaume**
We want to add support for comments. Our current strategy is the following:

Find the associated comment from the tree root. This gives us a way to retrieve leading and trailing comments from a Token. When we print the tree, recursively (in TreePrinter), we query tokens from the tree. For example for if:

```scala
case t: Term.If =>
  def body(expr: Term) = expr match {
    case b: Term.Block => space + dBlock(b.stats)
    case _: Term.If => space + print(expr)
    case _ => (line + print(expr)).nested(2)
  }
  val delse =
    if (t.elsep.tokens.isEmpty) empty
    else line + t.`else` + body(t.elsep)

  (
    t.`if` + space + t.`(` + print(t.cond) + t.`)` +
    body(t.thenp) +
    delse

  ).grouped
```

### CODE REVIEW

File: format/src/main/scala/org/scalafmt/internal/SyntaxTokens.scala [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+import scala.meta.tokens.Token._
+
+object SyntaxTokens {
+  private implicit class Util[A <: Tree](val tree: A) extends AnyVal {
+    def findAfter(f: Token => Boolean, p: A => Tree): Option[Token] = {
+      val end = p(tree).tokens.last
+      tree.tokens.dropWhile(_ != end).drop(1).find(f)
+    }
+  }
+  implicit class TermAssignSyntax(val tree: Term.Assign) extends AnyVal {
```
Comment (olafurpg) [inline format/src/main/scala/org/scalafmt/internal/SyntaxTokens.scala]: To follow scalameta/scalafix conventions, can we please use `implicit class XtensionTermAssignSyntax` as a naming convention for implicit classes.

File: format/src/main/scala/org/scalafmt/internal/SyntaxTokens.scala [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+    def `=`: Token = tree.findAfter(_.is[Equals], _.lhs).get
```
Comment (olafurpg) [inline format/src/main/scala/org/scalafmt/internal/SyntaxTokens.scala]: Let's use non-symbolic names for these methods and name them the same as the scala.meta.Token classes. In the old days scalameta tokens were actually named 

```scala
Token.`=`
Token.`=>`
...
```

However, this approach has several drawbacks

* hard to pronounce
* bad tooling support
  * breaks syntax highlighting
  * "go to matching curly" in many editors
  * "Go to class" doesn't work in IntelliJ
  * auto-completions suck

We bit the bullet and converted to non-symbolic names even if this mean having annoying names like `KwNew`. Sure maybe "tooling will improve" but I'm not holding my breadth.

File: format/src/main/scala/org/scalafmt/internal/SyntaxTokens.scala [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+
+object SyntaxTokens {
+  private implicit class Util[A <: Tree](val tree: A) extends AnyVal {
+    def findAfter(f: Token => Boolean, p: A => Tree): Option[Token] = {
+      val end = p(tree).tokens.last
+      tree.tokens.dropWhile(_ != end).drop(1).find(f)
+    }
+  }
+  implicit class TermAssignSyntax(val tree: Term.Assign) extends AnyVal {
+    def `=` : Token = tree.findAfter(_.is[Equals], _.lhs).get
```
Comment (olafurpg) [inline format/src/main/scala/org/scalafmt/internal/SyntaxTokens.scala]: `def tokensEqual: Token.Equals`

Let's use non-symbolic names for these methods and name them the same as the scala.meta.Token classes. In the old days scalameta tokens were actually named 

```scala
Token.`=`
Token.`=>`
...
```

However, this approach has several drawbacks

* hard to pronounce
* bad tooling support
  * breaks syntax highlighting
  * "go to matching curly" in many editors
  * "Go to class" doesn't work in IntelliJ
  * auto-completions suck

We bit the bullet and converted to non-symbolic names even if this mean having annoying names like `KwNew`. Sure maybe "tooling will improve" but I'm not holding my breadth.

File: format/src/main/scala/org/scalafmt/internal/SyntaxTokens.scala [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+  }
+
+  implicit class TermThisSyntax(val tree: Term.This) extends AnyVal {
+    def `.` : Option[Token] = {
```
Comment (olafurpg) [inline format/src/main/scala/org/scalafmt/internal/SyntaxTokens.scala]: `tokensDot: Option[Token.Dot]`

File: format/src/main/scala/org/scalafmt/internal/SyntaxTokens.scala [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+      tree.qual match {
+        case _: Name.Anonymous => None
+        case _ => tree.findAfter(_.is[Dot], _.qual)
+      }
+    }
+  }
+  implicit class TermSuperSyntax(val tree: Term.Super) extends AnyVal {
+    def `.` : Option[Token] = {
```
Comment (olafurpg) [inline format/src/main/scala/org/scalafmt/internal/SyntaxTokens.scala]: `tokensDot: Option[Token.Dot]`

File: format/src/main/scala/org/scalafmt/internal/SyntaxTokens.scala [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+      tree.thisp match {
+        case _: Name.Anonymous => None
+        case _ => tree.tokens.find(_.is[Dot])
+      }
+    }
+  }
+  implicit class TermSelectSyntax(val tree: Term.Select) extends AnyVal {
+    def `.` : Token = tree.findAfter(_.is[Dot], _.qual).get
```
Comment (olafurpg) [inline format/src/main/scala/org/scalafmt/internal/SyntaxTokens.scala]: `tokensDot: Token.Dot`

File: format/src/main/scala/org/scalafmt/internal/SyntaxTokens.scala [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+  }
+  implicit class TermInterpolate(val tree: Term.Interpolate) extends AnyVal {
+    def startQuote: Token = tree.tokens.find(_.is[Interpolation.Start]).get
+    def endQuote: Token = tree.tokens.find(_.is[Interpolation.End]).get
+  }
+
+  implicit class TermReturn(val tree: Term.Return) extends AnyVal {
+    def `return`: Token = tree.tokens.find(_.is[KwReturn]).get
```
Comment (olafurpg) [inline format/src/main/scala/org/scalafmt/internal/SyntaxTokens.scala]: `tokensReturn: Token.KwReturn`

File: format/src/main/scala/org/scalafmt/internal/SyntaxTokens.scala [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+  }
+
+  implicit class TermRepeated(val tree: Term.Repeated) extends AnyVal {
+    def `:` : Token = tree.findAfter(_.is[Colon], _.expr).get
```
Comment (olafurpg) [inline format/src/main/scala/org/scalafmt/internal/SyntaxTokens.scala]: `tokensColon: Token.Colon`

File: format/src/main/scala/org/scalafmt/internal/SyntaxTokens.scala [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+    def `_` : Token = tree.findAfter(_.is[Underscore], _.expr).get
```
Comment (olafurpg) [inline format/src/main/scala/org/scalafmt/internal/SyntaxTokens.scala]: etc., same applied to all the other signatures

File: format/src/main/scala/org/scalafmt/internal/SyntaxTokens.scala [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+  private def blockStartBrace(tree: Tree): Token =
+    tree.tokens.find(_.is[LeftBrace]).get
+  private def blockEndBrace[T <: Tree](
+      tree: T
+  )(childs: T => List[Tree]): Token = {
+    val stats = childs(tree)
+    if (stats.isEmpty) tree.tokens.find(_.is[RightBrace]).get
+    else tree.findAfter(_.is[RightBrace], _ => stats.last).get
+  }
+  private def `is*`(token: Token): Boolean = token match {
```
Comment (olafurpg) [inline format/src/main/scala/org/scalafmt/internal/SyntaxTokens.scala]: `def isAsterisk`

Please no fancy backtick names

File: format/src/main/scala/org/scalafmt/internal/SyntaxTokens.scala [hunk 11]
```diff
@@ -0,0 +1,95 @@
...
+    def `(` : Token = tree.tokens.find(_.is[LeftParen]).get
+    def `)` : Token = tree.findAfter(_.is[RightParen], _.cond).get
+    def `else`: Option[Token] = tree.findAfter(_.is[KwElse], _.thenp)
+  }
+  implicit class TermFor(val tree: Term.For) extends AnyVal {
+    def `for`: Token = tree.tokens.find(_.is[KwFor]).get
+  }
+  implicit class TermForYield(val tree: Term.ForYield) extends AnyVal {
+    def `for`: Token = tree.tokens.find(_.is[KwFor]).get
+    def `yield`: Token = tree.findAfter(_.is[KwYield], _.enums.last).get
```
Comment (olafurpg) [inline format/src/main/scala/org/scalafmt/internal/SyntaxTokens.scala]: why not find before the body?

**Review by olafurpg — COMMENTED** [review]
Very excited to see this come to life!

**Comment by olafurpg** [comment]
The `tokens*` prefix convention is up for debate, I think it has some nice benefits 

* similar to `.tokens` except only more precise
* auto-completion friendly, easy to discover all available green nodes of a type

File: format/src/main/scala/org/scalafmt/internal/SyntaxTokens.scala (hunk 3 above)
Comment (ShaneDelmore) [reply format/src/main/scala/org/scalafmt/internal/SyntaxTokens.scala]: I wouldn’t say you’re just holding your breath when it comes to improving tooling 😁

File: format/src/main/scala/org/scalafmt/internal/SyntaxTokens.scala [hunk 12]
```diff
@@ -3,92 +3,119 @@ package org.scalafmt.internal
...
+    def find[T <: Token](implicit ev: Classifier[Token, T]): Option[T] = {
+      find0[T](tree.tokens)(ev)
+    }
+    def findAfter[T <: Token](
+        p: A => Tree
+    )(implicit ev: Classifier[Token, T]): Option[T] = {
+      val tokens = after(p)
+      find0[T](tokens)(ev)
+    }
+    def findAfter2(f: Token => Boolean, p: A => Tree): Option[Token] = {
```
Comment (olafurpg) [inline format/src/main/scala/org/scalafmt/internal/SyntaxTokens.scala]: Can we use more self-descriptive names? overloaded `find` a `findAfter` would be better than `find0` and `findAfter2`

File: format/src/main/scala/org/scalafmt/internal/SyntaxTokens.scala [hunk 13]
```diff
@@ -3,92 +3,119 @@ package org.scalafmt.internal
...
+      val tokens = after(p)
+      tokens.find(f)
+    }
+    private def after(p: A => Tree): Tokens = {
       val end = p(tree).tokens.last
-      tree.tokens.dropWhile(_ != end).drop(1).find(f)
+      tree.tokens.dropWhile(_ != end).drop(1)
```
Comment (olafurpg) [inline format/src/main/scala/org/scalafmt/internal/SyntaxTokens.scala]: Can we binary search the index here?

```scala
  def binarySearch(tokens: Tokens, token: Token): Option[Int] = {
    def loop(lo: Int, hi: Int): Option[Int] = {
      if (lo > hi) None
      else {
        val mid = lo + ((hi - lo) / 2)
        val guess = tokens(mid)
        if (guess == token) Some(mid)
        else if (guess.end < token.end) loop(mid + 1, hi)
        else loop(lo, mid - 1)
      }
    }
    loop(0, tokens.length - 1)
  }
```

**Review by olafurpg — COMMENTED** [review]
Thanks for addressing the comments, only a few more

File: format/src/main/scala/org/scalafmt/internal/SyntaxTokens.scala (hunk 12 above)
Comment (MasseGuillaume) [reply format/src/main/scala/org/scalafmt/internal/SyntaxTokens.scala]: done.

File: format/src/main/scala/org/scalafmt/internal/SyntaxTokens.scala (hunk 13 above)
Comment (MasseGuillaume) [reply format/src/main/scala/org/scalafmt/internal/SyntaxTokens.scala]: Good idea :+1:

Comment (MasseGuillaume) [reply format/src/main/scala/org/scalafmt/internal/SyntaxTokens.scala]: Tokens extends IndexedSeq. Do you think that `indexOf` is implemented with a binary search?

Comment (olafurpg) [reply format/src/main/scala/org/scalafmt/internal/SyntaxTokens.scala]: IndexedSeq does not imply OrderedSeq, binary search is only possible if you know the data is ordered, which is the case for `Tokens`.

Comment (MasseGuillaume) [reply format/src/main/scala/org/scalafmt/internal/SyntaxTokens.scala]: Make sense :+1:. We should make Tokens OrderedSeq then?

Comment (olafurpg) [reply format/src/main/scala/org/scalafmt/internal/SyntaxTokens.scala]: There is no OrderedSeq, it's best to use our own binary search

**Review by olafurpg — APPROVED** [review]
LGTM 👍 This is a great start! I am also eager to use SyntaxTokens for improving the scalafix token patch api!

**Comment by MasseGuillaume** [comment]
@olafurpg I still have a long way to go. I would say 25%.

I'm doing every case of [CoverageSuite](https://github.com/scalacenter/scala-syntax/blob/master/tests/unit/src/test/scala/org/scalafmt/tests/CoverageSuite.scala).

**Pull request merged**